internal/rate/limiter.go

```diff
@@ -0,0 +1,114 @@
+package rate
+
+import (
+	"container/list"
+	"context"
+	"sync"
+	"time"
+
+	"golang.org/x/time/rate"
+)
+
+type (
+	limiterPerID struct {
+		evictList *list.List
+		items     map[string]*list.Element
+
+		lock sync.Mutex
+		t    timer
+		c    configInt
+	}
+
+	entry struct {
+		id       string
+		limiter  *rate.Limiter
+		lastCall time.Time
+	}
+)
+
+// NewLimiterPerID creates a limit limiter based on ID
+func NewLimiterPerID(rate rate.Limit, burst int, c *Config) *limiterPerID {
+	lpi := &limiterPerID{
+		evictList: list.New(),
+		items:     make(map[string]*list.Element),
+		t:         timerImpl{},
+		c:         toConfigInt(rate, burst, c),
+	}
+	go lpi.maintainLoop()
```

Comment:
every single IP will have a go routine to maintain it? This could lead to blowup of go routine right? if there is lots of ips. How about just use a single go routine to maintain the whole pool of ips?

Reply:
Nope. This is exactly the implementation you are saying. There will only be one maintain loop for handling http server rates.

Reply:
I see. looks good.